packages/backend/server/src/core/auth/job.ts

```diff
@@ -1,13 +1,35 @@
 import { Injectable } from '@nestjs/common';
 import { Cron, CronExpression } from '@nestjs/schedule';
 
+import { JobQueue, OnJob } from '../../base';
 import { Models } from '../../models';
 
+declare global {
+  interface Jobs {
+    'nightly.cleanExpiredUserSessions': {};
+  }
+}
+
 @Injectable()
 export class AuthCronJob {
-  constructor(private readonly models: Models) {}
+  constructor(
+    private readonly models: Models,
+    private readonly queue: JobQueue
+  ) {}
 
   @Cron(CronExpression.EVERY_DAY_AT_MIDNIGHT)
+  async nightlyJob() {
+    await this.queue.add(
+      'nightly.cleanExpiredUserSessions',
+      {},
+      {
+        // avoid duplicated jobs
+        jobId: 'nightly-auth-clean-expired-user-sessions',
```

Comment:
Interesting, it can actually implement a global task lock this way[![clap](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/WjlTemxb6oA4PgZFmj08/10e8450b-40ef-48ee-a846-327aa2dee627/clap.gif)](https://app.graphite.dev//settings/meme-library?org=toeverything)

Reply:
here is a cooler one

```ts
function sendAggregatedCommentNotificationEmail(comment: {}) {
  // 'you got 3 comments in workspace xxx'
  const job = await queue.remove('sendCommentsNotificationEmail', 'some-id')
  const comments = job?.data.comments ?? []
  comments.push(comment)
  
  // 'you got 4 comments in workspace xxx'
  await queue.add('sendCommentsNotificationEmail', 'some-id', { delay: 10m })
  //                                                                                                            ^ group notifications in 10 mins
}
```

Reply:
There is a risk of losing comment jobs; if the process crashes or the application restarts during the time between removal and the next addition, data will be lost.

However, for us, it is good enough for now.

Reply:
lose such notification is not fatal